### PR TITLE
[src] new feature: single task nodes

### PR DESCRIPTION
**Pull request by Lironrad**

### CODE REVIEW

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 1]
```diff
@@ -45,6 +47,8 @@ public abstract class BaseSpotinstCloud extends Cloud {
     private   ToolLocationNodeProperty          toolLocations;
     private   Boolean                           shouldUseWebsocket;
     private   Boolean                           shouldRetriggerBuilds;
+    private   Boolean                           isOneOffNodesEnabled =
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: Can we move to ctor? ehy do we need it?, consider change the terminology to "SingleTaskNode"

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 2]
```diff
@@ -587,25 +591,31 @@ protected Integer getNumOfExecutors(String instanceType) {
...
-            retVal = globalOverrideExecutorsNumber;
+        if (isOneOffNodesEnabled) {
+            retVal = 1;
         }
         else {
-            Integer defaultNumberOfExecutors = getDefaultExecutorsNumber(instanceType);
-
-            if (defaultNumberOfExecutors != null) {
-                retVal = defaultNumberOfExecutors;
+            if (globalOverrideExecutorsNumber != null) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: AwsSpotinstCloud should not override it any more, instead override other method that called : getOverridedExecuter which return const NO_OVERRIDE = -1, in aws it return the overrided value or -1 if no override

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 3]
```diff
@@ -587,25 +591,31 @@ protected Integer getNumOfExecutors(String instanceType) {
         Integer retVal;
 
         Integer globalOverrideExecutorsNumber = getExecutorsFromGlobalOverride();
+        boolean isOneOffNodesEnabled          = getIsOneOffNodesEnabled();
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: getExecutorsFromGlobalOverride => should be inside the else anyway

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 3 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 2 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 1 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 4]
```diff
@@ -585,35 +588,55 @@ protected PendingExecutorsCounts getPendingExecutors(ProvisionRequest request) {
...
             }
+
         }
 
         LOGGER.debug(String.format("instance type executors number was set to %s", retVal));
 
         return retVal;
     }
 
+    protected Integer getNumOfExecutorsOverride(String instanceType) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: rename to getOverridedNumberOfExecutors

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 5]
```diff
@@ -30,25 +32,27 @@ public abstract class BaseSpotinstCloud extends Cloud {
...
-    private   String                            vmargs;
-    private   EnvironmentVariablesNodeProperty  environmentVariables;
-    private   ToolLocationNodeProperty          toolLocations;
-    private   Boolean                           shouldUseWebsocket;
-    private   Boolean                           shouldRetriggerBuilds;
-    private   ComputerConnector                 computerConnector;
-    private   ConnectionMethodEnum              connectionMethod;
-    private   Boolean                           shouldUsePrivateIp;
-    private   SpotGlobalExecutorOverride        globalExecutorOverride;
+    protected static final Integer                           NO_OVERRIDE_NUM_OF_EXECUTORS = -1;
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: rename to NO_OVERRIDED_NUM_OF_EXECUTORS

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 6]
```diff
@@ -585,35 +588,55 @@ protected PendingExecutorsCounts getPendingExecutors(ProvisionRequest request) {
...
-        if (globalOverrideExecutorsNumber != null) {
-            LOGGER.debug(String.format("Overriding executors for instance type %s to be %s", instanceType,
-                                       globalOverrideExecutorsNumber));
-            retVal = globalOverrideExecutorsNumber;
+        if (isSingleTaskNodesEnabled) {
+            retVal = 1;
         }
         else {
-            Integer defaultNumberOfExecutors = getDefaultExecutorsNumber(instanceType);
+            Integer numOfExecutorsOverride = getNumOfExecutorsOverride(instanceType);
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: rename numOfExecutorsOverride to overridedNumOfExecutors

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 7]
```diff
@@ -585,35 +588,55 @@ protected PendingExecutorsCounts getPendingExecutors(ProvisionRequest request) {
...
+            boolean isNumOfExecutorsOverrideIsEnabled =
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: rename isNumOfExecutorsOverrideIsEnabled to isNumOfExecutorsOverrided

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 4 above)
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: change from Integer to integer

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java [hunk 8]
```diff
@@ -734,6 +757,29 @@ public SpotGlobalExecutorOverride getGlobalExecutorOverride() {
...
+        if (this.isSingleTaskNodesEnabled == null) {
+            return false;
+        }
+        else {
+            return isSingleTaskNodesEnabled;
+        }
+    }
+
+    @DataBoundSetter
+    public void setIsSingleTaskNodesEnabled(@CheckForNull Boolean isSingleTaskNodesEnabled) {
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: check (maybe with Shibel) what the annotation checkForNull does

File: src/main/java/hudson/plugins/spotinst/slave/SpotinstNonLocalizable.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.util.Locale;
+
+class SpotinstNonLocalizable extends NonLocalizable {
+
+    public SpotinstNonLocalizable(String nonLocalizable) {
+        super(nonLocalizable);
+    }
+
+    @Override
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/slave/SpotinstNonLocalizable.java]: Why do we need that override?

File: src/main/java/hudson/plugins/spotinst/slave/SpotinstNonLocalizable.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    public String toString(Locale locale) {
+        return super.toString(locale);
+    }
+
+    @Override
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/slave/SpotinstNonLocalizable.java]: Why do we need that override?, it does the same as super

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/slave/SpotinstComputer.java [hunk 11]
```diff
@@ -15,7 +20,44 @@
...
+            if (spotinstCloud != null) {
+                if (spotinstCloud.getIsSingleTaskNodesEnabled()) {
+                    String msg = String.format(
+                            "Node %s has accepted a job and 'Single Task Nodes' setting on Cloud %s is on. Node will not accept any more jobs.",
+                            spotinstNode.getNodeName(), spotinstCloud.getDisplayName());
+                    LOGGER.info(msg);
+                    this.setAcceptingTasks(false);
+                    // I see much better responsiveness from Jenkins in case of setting offline, both have same effect
+                    // should also be decided with UI because each combo looks different in the node list.
+                    this.setTemporarilyOffline(true, new SpotinstSingleTaskOfflineCause(new SpotinstNonLocalizable(msg)));
```
Comment (zivmessing) [inline src/main/java/hudson/plugins/spotinst/slave/SpotinstComputer.java]: extract to variable

**Review by zivmessing — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 4 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 5 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 6 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 7 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 4 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: done

File: src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java (hunk 8 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/cloud/BaseSpotinstCloud.java]: not necessary , deleted

File: src/main/java/hudson/plugins/spotinst/slave/SpotinstNonLocalizable.java (hunk 9 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/slave/SpotinstNonLocalizable.java]: not necessary , deleted

File: src/main/java/hudson/plugins/spotinst/slave/SpotinstNonLocalizable.java (hunk 10 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/slave/SpotinstNonLocalizable.java]: not necessary , deleted

File: src/main/java/hudson/plugins/spotinst/slave/SpotinstComputer.java (hunk 11 above)
Comment (Lironrad) [reply src/main/java/hudson/plugins/spotinst/slave/SpotinstComputer.java]: done

**Review by zivmessing — APPROVED** [review]

**Review by adivos — APPROVED** [review]

**Pull request merged**